### ISSUE TITLE
strange segment size in vod mapped mode

### ISSUE THREAD

**Issue by whatvn** [kaltura/nginx-vod-module#725]
Hello, 
We got strange issue with vod mapped mode, mapped json response look like this: 

```json
{  
   "playlistType":"vod",
   "discontinuity":true,
   "durations":[  
      120000,
      3597000,
      3480000
   ],
   "sequences":[  
      {  
         "clips":[  
            {  
               "type":"source",
               "clipFrom":3474000,
               "path":"/data/records/2017-11-05/34-1.stream/20-00.mp4"
            },
            {  
               "type":"source",
               "clipFrom":0,
               "path":"/data/records/2017-11-05/34-1.stream/21-00.mp4"
            },
            {  
               "type":"source",
               "clipFrom":0,
               "path":"/data/records/2017-11-05/34-1.stream/22-00.mp4"
            }
         ]
      }
   ]
}
```
m3u8 file is returned ok and file is play well until it reach segment 472, and 473. Segment 473 is set in playlist m3u8 with duration 9000 milisecs, but it size is only 376 bytes returns from vod module, that made video player start looping download again and again segment 472, and 473, cause segment loop. 

I attach link to mp4 file and index.m3u8 below. 
Here is vod configuration: 
```properties 
    vod hls;
    vod_align_segments_to_key_frames on;
    vod_hls_mpegts_interleave_frames on;
    vod_manifest_segment_durations_mode accurate;
    vod_segment_duration 9000;
    vod_segments_base_url "";
```
[index.m3u8](http://50.7.136.90:8082/download-xyz/index.m3u8)
[20-00.mp4](http://50.7.136.90:8082/download-xyz/20-00.mp4)
[21-00.mp4](http://50.7.136.90:8082/download-xyz/21-00.mp4)
[22-00.mp4](http://50.7.136.90:8082/download-xyz/22-00.mp4)

I personally think this is a bug in vod-module, or if there's something in configuration I can change to make this bug disappeared, please let me know. 

Thanks,

**Comment by erankor** [comment]
It's a problem in the mp4 - there is some video frame there with a duration of 28 seconds - 
```
ffprobe.sh -i /tmp/22-00.mp4 -show_packets -select_streams v
...
[PACKET]
codec_type=video
stream_index=0
pts=46974999
pts_time=521.944433
dts=46974999
dts_time=521.944433
duration=2519910
duration_time=27.999000
convergence_duration=N/A
convergence_duration_time=N/A
size=65489
pos=123773513
flags=K_
[/PACKET]
...
```
This segment falls somewhere within this frame, so it doesn't have any frames to output and returns empty.

**Comment by whatvn** [comment]
@erankor so is there's any way to skip error frame in first place instead of return error segment? 
thank,

**Comment by erankor** [comment]
It's not an error segment, it's a segment without frames, some players are ok with it, some are not.
But anyway, there can be many types of corruptions in the input files, I have no intention to try to solve them on my side.

**Comment by whatvn** [comment]
well, I am just curious about the way m3u8 is generated. 
in m3u8 playlist, this segment has duration 9 seconds, but the segment response itself is empty.
So is there anyway to report exactly segment duration inside m3u8 playlist?

**Comment by erankor** [comment]
With single video, you should get the correct duration, since you enabled 'accurate' mode. With playlist, I don't do it because it can be a heavy task to process all the metadata of all videos.

**Comment by whatvn** [comment]
@erankor: if i want to get actual segment duration, which part of source code should I look at? Since this problem happens to use so often, we should figure out the way to fix it. 
Thank

**Comment by erankor** [comment]
The code that handles the segmentation is in segmenter.c, specifically the function that calculates the accurate durations is segmenter_get_segment_durations_accurate. Currently, it doesn't support playlist, as I wrote above.

**Comment by whatvn** [comment]
@erankor : we got exactly this problem even with file "looks" normal to me. 
Json mapping: 
```
{
"playlistType": "vod",
"discontinuity": true,
"durations": [
2398000,
3603000,
2700000
],
"sequences": [
{
"clips": [
{
"type": "source",
"clipFrom": 1197000,
"path": "/data/records/2017-11-21/73-1.stream/13-00.mp4"
},
{
"type": "source",
"clipFrom": 0,
"path": "/data/records/2017-11-21/73-1.stream/14-00.mp4"
},
{
"type": "source",
"clipFrom": 0,
"path": "/data/records/2017-11-21/73-1.stream/15-00.mp4"
}
]
}
]
}
```

[13-00.mp4](http://109.236.90.139:8082/debug/13-00.mp4)
[14-00.mp4](http://109.236.90.139:8082/debug/14-00.mp4)
[15-00.mp4](http://109.236.90.139:8082/debug/15-00.mp4)

segment 375 and 376 cause error which has size 692kb and 376 bytes. 

Those file doesn't have frame duration with abnormal duration, but still cause error with empty segment. 

Can you take a look?

**Comment by erankor** [comment]
It's the same issue as before, but for some reason here ffprobe does not output the actual duration - 
```
ffprobe.sh -i /tmp/14-00.mp4 -show_packets -select_streams v
...
[PACKET]
codec_type=video
stream_index=0
pts=87213600
pts_time=969.040000
dts=87210000
dts_time=969.000000
duration=3600
duration_time=0.040000
convergence_duration=N/A
convergence_duration_time=N/A
size=58457
pos=232014054
flags=K_
[/PACKET]
[PACKET]
codec_type=video
stream_index=0
pts=89103600
pts_time=990.040000
dts=89100000
dts_time=990.000000
duration=3600
duration_time=0.040000
convergence_duration=N/A
convergence_duration_time=N/A
size=65184
pos=232115268
flags=K_
[/PACKET]
...
```
These are two adjacent frames in the ffprobe output, and while the duration is reported as 0.04, the difference in timestamps (both pts & dts) is 21 seconds.

**Comment by whatvn** [comment]
If you could bring the option to reporter exact segment duration and it can be switch on of off, it would be great with us.

Because it will cause performance problem, user can choose if it should be on or off.

--
Hưng

**Comment by whatvn** [comment]
hello @erankor ,

it's me again. Technically, instead of return empty segment in this case, can we return a "playable segment" for the player?

**Comment by erankor** [comment]
That would probably be easier to implement, you can have it redirect to the next segment when this happens. However, this sounds like a bit of an odd feature (the same content will play multiple times), I wouldn't want to merge it here, you can patch it locally...

**Comment by whatvn** [comment]
@erankor yes that's pretty easy to do so, if i do it i will write different module to do instead of changing vod module code base. 

But what i actually want is instead of redirect (internal) request segment to its next segment, I want to return a "not so empty segment", a playable plain frame to there's no repeat content. is it possible?
